### PR TITLE
Server-streaming method support in gRPC client pseudo-Node

### PR DESCRIPTION
**Pull request by daviddrysdale**
# Checklist

- [x] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [x] I have written tests that cover the code changes.
  - [x] I have checked that these tests are run by [Cloudbuild](cloudbuild.yaml)
  - [ ] I have updated [documentation](docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [ ] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

File: oak/server/rust/oak_runtime/src/node/grpc/client.rs [hunk 1]
```diff
@@ -63,67 +67,134 @@ impl GrpcClientNode {
...
+            error!(
+                "Couldn't read gRPC request from the invocation: {:?}",
                 error
+            );
+            error
+        })?;
+        debug!("Incoming gRPC request: {:?}", request);
+
+        // Connect to an external gRPC service.
+        let mut grpc_client = self.connect().await.map_err(|error| {
```
Comment (ipetr0v) [inline oak/server/rust/oak_runtime/src/node/grpc/client.rs:112]: Does it mean that a gRPC client will create a new connection for each invocation?

File: oak/server/rust/oak_runtime/src/node/grpc/client.rs [hunk 2]
```diff
@@ -63,67 +67,134 @@ impl GrpcClientNode {
...
+        let path = request.method_name.parse().map_err(|error| {
+            error!("Invalid URI {}: {}", request.method_name, error);
+            invocation.send_error(rpc::Code::InvalidArgument, "Invalid URI", runtime);
+            OakStatus::ErrInternal
+        })?;
+
+        // Forward the request to the external gRPC service and wait for the response(s).
+        let request = tonic::Request::new(request.req_msg);
+        let request_stream =
+            request.map(|m| futures_util::stream::once(futures_util::future::ready(m)));
```
Comment (ipetr0v) [inline oak/server/rust/oak_runtime/src/node/grpc/client.rs:133]: Will it send only a single request in this stream?

File: oak/server/rust/oak_runtime/src/node/grpc/client.rs [hunk 3]
```diff
@@ -63,67 +67,134 @@ impl GrpcClientNode {
...
+                error!("Failed to read response: {}", error);
+                invocation.send_error(rpc::Code::Internal, "Failed to read response", runtime);
+                OakStatus::ErrInternal
+            })?;
+
+            if let Some(message) = message {
+                let encap_rsp = GrpcResponse {
+                    rsp_msg: message,
+                    status: None,
+                    last: false,
```
Comment (ipetr0v) [inline oak/server/rust/oak_runtime/src/node/grpc/client.rs:155]: Will Oak streaming methods work if there will not be a `last: true` message?

File: oak/server/rust/oak_runtime/src/node/grpc/client.rs (hunk 1 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/node/grpc/client.rs:112]: Yes, but I think that's also how it behaved before (every invocation of `GrpcRequestHandler::unary_request()` would call `GrpcRequestHandler::connect()` which would overwrite `self.dispatcher`).

I did have a quick attempt at making the connection persistent between method invocations but it didn't immediately work, so I'll add an issue or TODO to investigate further.

File: oak/server/rust/oak_runtime/src/node/grpc/client.rs (hunk 2 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/node/grpc/client.rs:133]: Yes, this is only adding support for server-streaming methods, not client-streaming.  I've tried to structure things to make it easier to possibly add client-streaming later, but I don't yet know how or if the async stuff will work (the code fundamentally needs to be waiting on two disparate work sources at the same time, the `wait_on_channels` for Node-triggered requests and the HTTP connection for gRPC-traffic-triggered responses).

File: oak/server/rust/oak_runtime/src/node/grpc/client.rs (hunk 3 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/node/grpc/client.rs:155]: They should treat the channel closure as an indication that the last message in the channel was `last: true`; I believe the [C++ code behaved the same](https://github.com/project-oak/oak/blob/2d79f622b622848fd25aa93a1edce672be3bf8c1/oak/server/grpc_client_node.cc#L125)

**Pull request merged**